lib/Producer.js

```diff
@@ -0,0 +1,192 @@
+      let addresses = config.lookupdHTTPAddresses;
+      if (!Array.isArray(addresses)) {
+        addresses = [addresses];
+      }
+      this.lookupd = normalizeAddress(addresses);
+      this.counter = 0;
+      this.strategy = (options && options.strategy) || Producer.ROUND_ROBIN;
+    }
+  }
+
+  connect(callback) {
+    this.conns = [];
+    this._closed = false;
+    if (this.lookupd) {
+      return nsqlookup(this.lookupd).then((nodes) => {
+        return Promise.map(nodes, (node) => {
+          return this.connectNsqd(node.broadcast_address, node.tcp_port, this.opts);
+        }).then((error) => {
+          return this.conns;
+        }, (error) => {
+          debug('connect to nsqd failed: %s', error);
+          throw error;
+        });
+      }, (errors) => {
+        debug('lookup errors %j', errors);
+        throw errors;
+      }).asCallback(callback);
+    }
+
+    //connect directly
+    return this.connectNsqd(this.nsqd.host, this.nsqd.port, this.opts).then(() => {
+      return this.conns;
+    }).asCallback(callback);
+  }
+
+  /**
+   * Connect to an NSQD and put it in the pool.
+   *
+   * @private
+   */
+  connectNsqd(host, port, options) {
+    return new Promise((resolve, reject) => {
+      const writer = Promise.promisifyAll(new nsq.Writer(host, port, options));
+      writer.connect();
+      writer.on('error', reject);
+      writer.on('ready', () => {
+        debug(`writer on ready: ${writer.nsqdHost}:${writer.nsqdPort}`);
+        this.conns.push(writer);
+        resolve(writer);
+      });
+      writer.on('closed', () => {
+        this.reconnectNsqd(host, port, options);
+      });
+    });
+  }
+
+  /**
+   * Reconnect, if and only if the connection is in our pool.
+   *
+   * @private
+   */
+  reconnectNsqd(host, port, options) {
+    if (this._closed) {
+      return;
+    }
+    const idx = indexOfConnection(this.conns, host, port);
+    debug('on closed, idx of cached connection: %d', idx);
+    if (idx === -1) {
+      return;
+    }
+    this.conns.splice(idx, 1);
+    //TODO should be able to specify the retry strategy
+    promiseRetry((retry, number) => {
+      debug('connect attempts %d', number);
+      return this.connectNsqd(host, port, options).catch((err) => {
+        debug('retry when err: ' + err);
+        retry(err);
+      });
+    });
+  }
+
+  produce(topic, msg, options, callback) {
+    if (typeof options === 'function') {
+      callback = options;
+      options = {};
+    }
+
+    if (!this.conns || this.conns.length === 0) {
+      return Promise.reject(new Error('No connections to nsqd'))
+        .asCallback(callback);
+    }
+
+    if (options.retry && this.strategy === Producer.FAN_OUT) {
+      return Promise.reject(new Error('Retry on produce level is not supported in fanout strategy'))
+        .asCallback(callback);
+    }
+
+    if (options.retry) {
+      return Promise.resolve(promiseRetry({
+        retries: options.retry === true ? 10 : options.retry
```

Comment:
The logic is not same as before, here the whole option `options.retry` would be set as `retries`, which means callers lost the capability of setting other fields(factor, maxTimeout, etc.)